zkevm-circuits/src/pi_circuit.rs

```diff
@@ -0,0 +1,978 @@
+//! Public Input Circuit implementation
+
+use std::io::Cursor;
+use std::marker::PhantomData;
+
+use eth_types::geth_types::BlockConstants;
+use eth_types::H256;
+use eth_types::{geth_types::GethData, U256};
+use eth_types::{geth_types::Transaction, Address, Field, ToLittleEndian, ToScalar, Word};
+use ethers_core::types::Block;
+use halo2_proofs::arithmetic::BaseExt;
+use halo2_proofs::plonk::Instance;
+
+use crate::tx_circuit::sign_verify::SignData;
+use crate::tx_circuit::{tx_to_sign_data, TxFieldTag};
+use crate::util::random_linear_combine_word as rlc;
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter, Region, SimpleFloorPlanner},
+    plonk::{Advice, Circuit, Column, ConstraintSystem, Error, Fixed, Selector},
+    poly::Rotation,
+};
+
+/// Fixed by the spec
+const TX_LEN: usize = 9;
+const BLOCK_LEN: usize = 7 + 256;
+const EXTRA_LEN: usize = 2;
+
+/// Values of the block table (as in the spec)
+#[derive(Clone, Default, Debug)]
+pub struct BlockValues {
+    coinbase: Address,
+    gas_limit: Word,
+    number: u64,
+    timestamp: Word,
+    difficulty: Word,
+    base_fee: Word, // NOTE: BaseFee was added by EIP-1559 and is ignored in legacy headers.
+    chain_id: Word,
+    history_hashes: Vec<U256>,
```

Comment:
I think that made sense, but why are they of type `Vec<Word>` in `GethData`?  When in doubt I went for the type that was used there.

Reply:
I'm not sure why `history_hashes` uses `Vec<Word>` instead of `Vec<H256>`.  Maybe we can leave topic about choosing the best types between `Word`, `U256` and `H256` for later.

Reply:
I've changed all hashes to `H256` but I did a rather ugly conversion as there is no `From<U256` for `H256`. Maybe there is a reason for them to be `U256`...